libraries/AP_LocationDB/AP_LocationDB.h

```diff
@@ -0,0 +1,139 @@
+#pragma once
+
+#include "AP_LocationDB_config.h"
+
+#if AP_LOCATIONDB_ENABLED
+
+#include <AP_Common/AP_Common.h>
+#include <AP_Common/Location.h>
+#include <AP_Math/AP_Math.h>
+#include <AP_HAL/AP_HAL_Boards.h>
+#include <AP_HAL/Semaphores.h>
+#include <AP_Vehicle/AP_Vehicle_Type.h>
+#include <AC_Avoidance/AP_OABendyRuler.h>
+
+class AP_LocationDB_Item {
+public:
+    friend class AP_Filesystem_LocationDB;
+    friend class AP_LocationDB;
+
+    AP_LocationDB_Item() {};
+    AP_LocationDB_Item(uint32_t key, uint32_t timestamp_ms, Vector3f pos_cm_NEU, Vector3f vel_cm_NEU, Vector3f acc_cm_NEU, float heading_cdeg, float radius_cm, uint8_t populated_fields);
+
+    // to initialise an empty database item with given data
+    void init(uint32_t key, uint32_t timestamp_ms, Vector3f pos_cm_NEU, Vector3f vel_cm_NEU, Vector3f acc_cm_NEU, float heading_cdeg, float radius_cm, uint8_t populated_fields);
+    
+    uint32_t get_timestamp_ms() { return _timestamp_ms; }
+    uint32_t get_key() { return _key; }
+    bool get_pos_cm_NEU(Vector3f &ret) const;
+    bool get_vel_cm_NEU(Vector3f &ret) const;
+    bool get_acc_cm_NEU(Vector3f &ret) const;
+    bool get_heading_cdeg(float &heading) const;
+    bool get_radius_cm(float &ret) const;
+
+    enum class DataField : uint8_t {
+        POS = (1U << 0),
+        VEL = (1U << 1),
+        ACC = (1U << 2),
+        HEADING = (1U << 3),
+        RADIUS = (1U << 4),
+    };
+
+    enum class Flag : uint8_t {
+        AVOID = (1U << 0),
+    };
+
+private:
+    uint32_t _key;
+    uint32_t _timestamp_ms;
+    Vector3f _pos;
+    Vector3f _vel;
+    Vector3f _acc;
+    float _heading;
+    float _radius;
+    uint8_t _flags;
+
+    uint8_t _populated_fields = 0;
+
+    bool is_field_populated(DataField option) const {
+        return (_populated_fields & (uint8_t)option) != 0;
+    }
+};
+
+class AP_LocationDB {
+public:
+    friend class AP_Filesystem_LocationDB;
+    friend class AP_OABendyRuler;
+
+    AP_LocationDB();
+
+    CLASS_NO_COPY(AP_LocationDB);
+
+    static AP_LocationDB *get_singleton() {
+        return _singleton;
+    }
+
+    void init();
+    void update();
+
+    enum class KeyDomain : uint8_t {
+        // 0 should not be used
+        MAVLINK = 1U,
+#if AP_LOCATIONDB_KEYDOMAIN_ADSB_ENABLED
+        ADSB = 2U,
+#endif
+#if AP_LOCATIONDB_KEYDOMAIN_SCRIPTING_ENABLED
+        SCRIPTING = 3U,
+#endif
+    };
+
+    bool add_item(AP_LocationDB_Item item); // add item to the location database
+    bool get_item(const uint32_t key, AP_LocationDB_Item &ret); // get item with given key from location database
```

Comment:
Doesn't modify the class
```suggestion
    bool get_item(const uint32_t key, AP_LocationDB_Item &ret) const;// get item with given key from location database
```

Reply:
I think we can't make this method const as it modifies the value of semaphore internally.

Reply:
Ah ok, worth a shot:)